akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/LocalReceptionist.scala

```diff
@@ -37,98 +38,206 @@ private[akka] object LocalReceptionist extends ReceptionistBehaviorProvider {
-  type SubscriptionsKV[K <: AbstractServiceKey] = ActorRef[ReceptionistMessages.Listing[K#Protocol]]
-  type SubscriptionRegistry = TypedMultiMap[AbstractServiceKey, SubscriptionsKV]
+  private type Service[K <: AbstractServiceKey] = ActorRef[K#Protocol]
+  private type Subscriber[K <: AbstractServiceKey] = ActorRef[ReceptionistMessages.Listing[K#Protocol]]
 
-  sealed trait InternalCommand
-  final case class RegisteredActorTerminated[T](key: ServiceKey[T], ref: ActorRef[T]) extends InternalCommand
-  final case class SubscriberTerminated[T](key: ServiceKey[T], ref: ActorRef[ReceptionistMessages.Listing[T]])
+  private sealed trait InternalCommand
+  private final case class RegisteredActorTerminated[T](ref: ActorRef[T]) extends InternalCommand
+  private final case class SubscriberTerminated[T](ref: ActorRef[ReceptionistMessages.Listing[T]])
       extends InternalCommand
 
-  override def behavior: Behavior[Command] = Behaviors.setup { ctx =>
-    ctx.setLoggerName(classOf[LocalReceptionist])
-    behavior(TypedMultiMap.empty[AbstractServiceKey, KV], TypedMultiMap.empty[AbstractServiceKey, SubscriptionsKV])
-      .narrow[Command]
+  private object State {
+    def empty =
+      State(
+        TypedMultiMap.empty[AbstractServiceKey, Service],
+        Map.empty,
+        TypedMultiMap.empty[AbstractServiceKey, Subscriber],
+        Map.empty)
   }
 
-  private def behavior(serviceRegistry: LocalServiceRegistry, subscriptions: SubscriptionRegistry): Behavior[Any] = {
-
-    // Helper to create new state
-    def next(
-        newRegistry: LocalServiceRegistry = serviceRegistry,
-        newSubscriptions: SubscriptionRegistry = subscriptions) =
-      behavior(newRegistry, newSubscriptions)
-
-    /*
-     * Hack to allow multiple termination notifications per target
-     * FIXME #26505: replace by simple map in our state
-     */
-    def watchWith(ctx: ActorContext[Any], target: ActorRef[_], msg: InternalCommand): Unit =
-      ctx.spawnAnonymous[Nothing](Behaviors.setup[Nothing] { innerCtx =>
-        innerCtx.watch(target)
-        Behaviors.receiveSignal[Nothing] {
-          case (_, Terminated(`target`)) =>
-            ctx.self ! msg
-            Behaviors.stopped
+  /**
+   * @param services current registered services per key
+   * @param servicesPerActor current registered service keys per actor (needed for unregistration since an actor can implement several services)
+   * @param subscriptions current subscriptions per service key
+   * @param subscriptionsPerActor current subscriptions per subscriber (needed since a subscriber can subscribe to several keys) FIXME is it really needed?
+   */
+  private final case class State(
+      services: TypedMultiMap[AbstractServiceKey, Service],
+      servicesPerActor: Map[ActorRef[_], Set[AbstractServiceKey]],
+      subscriptions: TypedMultiMap[AbstractServiceKey, Subscriber],
+      subscriptionsPerActor: Map[ActorRef[_], Set[AbstractServiceKey]]) {
```

Comment:
With a large number of `services, `subscriptions` and or actors thereof, with no bounds, wouldn't we see a problem in our processing.

Reply:
Yes, the receptionist will not work well for use cases with, say, millions of services and subscribers. I think it is fair to say that we cannot cover every use case and that for such a use case the user would have to abstract over it and only use it as the initial way to find the top level actor of their solution from the receptionist or something along those lines.

Reply:
Mental note to add some concept of bounds in the doc for massive use case, what that might constitute.

Reply:
We can do that directly in this PR I think